File: Java/Sorting/InsertionSort.java
/*The array is virtually split into a sorted and an unsorted part. Values from the unsorted part are picked and placed at the correct position in the sorted part. */
import java.util.*;

public class InsertionSort {
    static void insertionsort(int arr[]){
        for(int i=1;i<arr.length;i++){
            int temp = arr[i];
            int j = i-1; //for storing sorted elements

            while(j>=0 && arr[j]>temp){
                arr[j+1] = arr[j];
                j--;
            }
            arr[j+1] = temp;
        }
    }

    //printing array
    static void printArray(int arr[]){
        for(int i=0;i<arr.length;i++){
            System.out.print(arr[i]+ " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter size of array:");
        int n = sc.nextInt();
        int arr[] = new int[n];

        //array elements
        System.out.print("Enter array elements:");
        for(int i=0;i<n;i++){
        arr[i] = sc.nextInt();
        }
        insertionsort(arr);
        System.out.println("Sorted elements are ");
        printArray(arr);
    }
    
}

File: C Program/Sorting Technique/Insertion_Sort.c
#include<stdio.h>

void insertion_sort(int arr[],int n){
    for(int i=1;i<n;i++){
        int key = arr[i];
        int j = i-1;

        while(key < arr[j] && j>=0){
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = key;
    }
}

void print(int arr[],int n){
    for(int i=0;i<n;i++){
        printf("%d ",arr[i]);
    }
    printf("\n");
}

int main(){
    int n;
    printf("Enter size of an array:");
    scanf("%d",&n);

    int arr[n];
    printf("Enter array elements:");
    for(int i=0;i<n;i++){
        scanf("%d",&arr[i]);
    }

    insertion_sort(arr,n);

    printf("Sorted elements are:");
    print(arr,n);

}
